src/rgw/driver/d4n/d4n_policy.cc

```diff
@@ -1,401 +1,427 @@
+}
 
-      if (ret < 0)
-	return {};
-    } else if (std::stoi(localWeightStr) < minWeight) {
-      minWeight = std::stoi(localWeightStr);
-      victimName = it->key;
-    }
+int LFUDAPolicy::exist_key(std::string key) {
+  if (entries_map.count(key) != 0) {
+    return true;
   }
 
-  /* Get victim cache block */
-  CacheBlock victimBlock;
-  victimBlock.cacheObj.objName = victimName;
-  BlockDirectory blockDir;
-  blockDir.init(cct);
+  return false;
+}
 
-  int ret = blockDir.get_value(&victimBlock);
+#if 0
+int LFUDAPolicy::get_block(const DoutPrefixProvider* dpp, CacheBlock* block, rgw::cache::CacheDriver* cacheDriver, optional_yield y) {
+  response<std::string> resp;
+  int age = get_age(y);
 
-  if (ret < 0)
-    return {};
+  if (exist_key(build_index(block->cacheObj.bucketName, block->cacheObj.objName, block->blockID, block->size))) { /* Local copy */
+    auto it = entries_map.find(build_index(block->cacheObj.bucketName, block->cacheObj.objName, block->blockID, block->size));
+    it->second->localWeight += age;
+    return cacheDriver->set_attr(dpp, block->cacheObj.objName, "localWeight", std::to_string(it->second->localWeight), y);
+  } else {
+    if (eviction(dpp, block->size, cacheDriver, y) < 0)
+      return -1; // what if eviction turns into infinite loop? -Sam
 
-  return victimBlock;
+    int exists = dir->exist_key(block, y);
+    if (exists > 0) { /* Remote copy */
+      if (dir->get(block, y) < 0) {
+	return -1;
+      } else {
+	if (!block->hostsList.empty()) { 
+	  block->globalWeight += age;
+	  
+	  if (dir->update_field(block, "globalWeight", std::to_string(block->globalWeight), y) < 0) {
+	    return -1;
+	  } else {
+	    return 0;
+	  }
+	} else {
+          return -1;
+        }
+      }
+    } else if (!exists) { /* No remote copy */
+      // localWeight += age;
+      //return cacheDriver->set_attr(dpp, block->cacheObj.objName, "localWeight", std::to_string(it->second->localWeight), y);
+      return 0;
+    } else {
+      return -1;
+    }
+  }
 }
+#endif
 
-int LFUDAPolicy::get_block(const DoutPrefixProvider* dpp, CacheBlock* block, rgw::cache::CacheDriver* cacheNode) {
-  std::string key = "rgw-object:" + block->cacheObj.objName + ":directory";
-  std::string localWeightStr = cacheNode->get_attr(dpp, block->cacheObj.objName, "localWeight"); // change to block name eventually -Sam
-  int localWeight = -1;
+int LFUDAPolicy::eviction(const DoutPrefixProvider* dpp, uint64_t size, optional_yield y) {
+  uint64_t freeSpace = cacheDriver->get_free_space(dpp);
 
-  if (!client.is_connected())
-    find_client(dpp, &client);
+  while (freeSpace < size) {
+    CacheBlock victim = find_victim(dpp, y);
 
-  if (localWeightStr.empty()) { // figure out where to set local weight -Sam
-    int ret = cacheNode->set_attr(dpp, block->cacheObj.objName, "localWeight", std::to_string(get_age())); 
-    localWeight = get_age();
-
-    if (ret < 0)
+    if (victim.cacheObj.objName.empty()) {
```

Comment:
shouldn't this check do something with block id, since its a victim 'block' that we are checking for

Reply:
This is just to ensure a victim was retrieved successfully from the `find_victim` method. If `find_victim` fails, an empty initializer list is returned from the method, so I am simply checking if the block is empty or not. If there is a better way to do this, I can incorporate it and remove this line

Reply:
Block id being zero can be a valid block id correct? Is there any other attribute of the block that can be used here? If not, let this check be.

Reply:
Yes, blockID just represents the offset so it can be 0. I will update the `find_victim` method to return a `nullptr` upon failure.